SDALib-CPP/SDALibTests/SDADriverTest.cpp

```diff
@@ -16,87 +16,192 @@ void DriverSide()
+{
+	p_server.ReceiveDataAsync();
+	ASSERT_EQ(p_server.SendData("1", 1), IPCLIB_SUCCEED);
 
 	// return result
-	server.AwaitData(buffer, TEST_BUFFER_SIZE);
+	ASSERT_DURATION_LE(1, p_server.AwaitData(p_buffer, TEST_BUFFER_SIZE));
 	std::vector<std::string> actionData;
-	GetMsgVector(buffer, TEST_BUFFER_SIZE, actionData);
-	ASSERT_ALMOST_EQ(1, stof(actionData[0]));
-	ASSERT_ALMOST_EQ(2, stof(actionData[1]));
+	GetMsgVector(p_buffer, TEST_BUFFER_SIZE, actionData);
+	ASSERT_ALMOST_EQ(0.75, stof(actionData[0]));
+	ASSERT_ALMOST_EQ(0.5, stof(actionData[1]));
+	ASSERT_ALMOST_EQ(2, stof(actionData[2]));
+	ASSERT_EQ(1, stoi(actionData[3]));
+}
 
+/// @brief            Tests if the connection can be stopped, so that the AI can shutdown
+/// @param  p_server  The server that sends the message
+/// @param  p_buffer  The buffer to store/receive a message
+void TestStopConnection(ServerSocket& p_server, char* p_buffer)
+{
 	// stop the connection
-	ASSERT_EQ(server.SendData("STOP", 4), IPCLIB_SUCCEED);
-	server.AwaitData(buffer, TEST_BUFFER_SIZE);
-	ASSERT_TRUE(buffer[0] == 'O' && buffer[1] == 'K' && buffer[2] == '\0');
+	p_server.ReceiveDataAsync();
+	ASSERT_EQ(p_server.SendData("STOP", 4), IPCLIB_SUCCEED);
+	ASSERT_DURATION_LE(1, p_server.AwaitData(p_buffer, TEST_BUFFER_SIZE));
+	ASSERT_TRUE(p_buffer[0] == 'O' && p_buffer[1] == 'K' && p_buffer[2] == '\0');
 	std::this_thread::sleep_for(std::chrono::milliseconds(10)); // wait to disconnect client
-	server.CloseServer();
-}
\ No newline at end of file
+	p_server.CloseServer();
+}
+
+/// @brief				 The server side of the tests
+/// @param  p_server	 The server socket that has a connection to the driver
+/// @param  p_situations The amount of situations the driver has to handle
+void ServerSide(ServerSocket& p_server, int p_situations)
+{
+	ASSERT_DURATION_LE(3, while (!p_server.Connected()) {});
+	char buffer[TEST_BUFFER_SIZE];
+
+	TestActive(p_server, buffer);
+
+	TestOrder(p_server, buffer);
+
+	TestTestAmount(p_server, buffer);
+
+	TestSituation(p_server, buffer); // 1 test situation
+
+	for(int i = 0; i < p_situations; i++)
+	{
+		TestSituation(p_server, buffer);
+	}
+
+	TestStopConnection(p_server, buffer);
+}
+
+/// @brief				 Tests the behaviour of a driver when it does not have to wait on a connection of the server
+/// @param  p_situations The amount of different situations the driver has to handle
+void NoWaitTest(int p_situations)
+{
+	// connect server and black box
+	ServerSocket server;
+	ASSERT_EQ(server.Initialize(), IPCLIB_SUCCEED);
+	server.ConnectAsync();
+	std::thread t = std::thread(DriverSide);
+	t.detach();
+
+	ServerSide(server, p_situations);
+}
+
+/// @brief Tests a full session with a server
+TEST(DriverTests, NoWaitNoSituations)
+{
+	NoWaitTest(0);
+}
+
+/// @brief Tests a full session with a server
+TEST(DriverTests, NoWaitOneSituation)
+{
+	NoWaitTest(1);
+}
+
+/// @brief Tests a full session with a server
+TEST(DriverTests, NoWaitMultipleSituations)
+{
+	NoWaitTest(4);
+}
```

Comment:
Maybe add TEST_CASE() from the speed dreams TestUtils.h to this repository too and use that?